crates/federate/src/worker.rs

```diff
@@ -137,148 +215,245 @@ impl InstanceWorker {
+    if updated.add(Days::new(1)) < Utc::now() {
+      self.instance.updated = Some(Utc::now());
+
+      let form = InstanceForm::builder()
+        .domain(self.instance.domain.clone())
+        .updated(Some(naive_now()))
+        .build();
+      Instance::update(&mut self.pool(), self.instance.id, form).await?;
+    }
+    Ok(())
+  }
+  /// checks whether the highest successful id can be updated and writes to db if so
+  async fn pop_successfuls_and_write(
+    &mut self,
+    successfuls: &mut BinaryHeap<SendSuccessInfo>,
+    force_write: bool,
+  ) -> Result<()> {
+    let Some(mut last_id) = self.state.last_successful_id else {
+      tracing::warn!("should be impossible: last successful id is None");
+      return Ok(());
+    };
+    tracing::debug!(
+      "last: {:?}, next: {:?}, currently in successfuls: {:?}",
+      last_id,
+      successfuls.peek(),
+      successfuls.iter()
+    );
+    while successfuls
+      .peek()
+      .map(|a| &a.activity_id == &ActivityId(last_id.0 + 1))
+      .unwrap_or(false)
     {
-      id = ActivityId(id.0 + 1);
-      processed_activities += 1;
-      let Some(ele) = get_activity_cached(pool, id)
-        .await
-        .context("failed reading activity from db")?
-      else {
-        tracing::debug!("{}: {:?} does not exist", self.instance.domain, id);
-        self.state.last_successful_id = Some(id);
-        continue;
-      };
-      if let Err(e) = self.send_retry_loop(pool, &ele.0, &ele.1).await {
+      let next = successfuls.pop().unwrap();
+      last_id = next.activity_id;
+      self.state.last_successful_id = Some(next.activity_id);
+      self.state.last_successful_published_time = next.published;
+    }
+
+    let save_state_every = chrono::Duration::from_std(SAVE_STATE_EVERY_TIME).expect("not negative");
+    if force_write || (Utc::now() - self.last_state_insert) > save_state_every {
+      self.save_and_send_state().await?;
+    }
+    Ok(())
+  }
+
+  async fn spawn_send_if_needed(
+    &mut self,
+    activity_id: ActivityId,
+    report: UnboundedSender<SendActivityResult>,
+  ) -> Result<()> {
+    let Some(ele) = get_activity_cached(&mut self.pool(), activity_id)
+      .await
+      .context("failed reading activity from db")?
+    else {
+      tracing::debug!("{}: {:?} does not exist", self.instance.domain, activity_id);
+      report.send(SendActivityResult::Success(SendSuccessInfo {
+        activity_id,
+        published: None,
+        was_skipped: true,
+      }))?;
+      return Ok(());
+    };
+    let activity = &ele.0;
+    let inbox_urls = self
+      .get_inbox_urls(activity)
+      .await
+      .context("failed figuring out inbox urls")?;
+    if inbox_urls.is_empty() {
+      tracing::debug!("{}: {:?} no inboxes", self.instance.domain, activity.id);
+      report.send(SendActivityResult::Success(SendSuccessInfo {
+        activity_id,
+        published: Some(activity.published),
```

Comment:
I think this should be None as it didnt actually connect to the instance.

Reply:
I'm not sure, this value is shown in that status page I wrote ( https://phiresky.github.io/lemmy-federation-state/ ) as well as one of those grafana charts https://grafana.lem.rocks/d/edf3gjrxxnocgd/federation-health-activities-behind?orgId=1&var-instance=All&var-remote_instance=lemdro.id&var-remote_software=All&from=now-24h&to=now

The behaviour you want is how it currently works, but it can be a bit misleading because if you look at / chart the published time for federation from a large to a small instance that's only subscribed to a few small communities, then it will show the last published time as a days ago even though federation is up to date. 

I don't have a strong opinion one way or another though, so let me know.

Reply:
Yeah Im not sure either. Maybe just leave a coment that it may be changed to None.